internal/buffer/buffer.go

```diff
@@ -909,10 +912,22 @@ func (b *Buffer) UpdateRules() {
 		b.SyntaxDef = findRuntimeSyntaxDef(syntaxFile, header)
 	}
 
+	if b.SyntaxDef != nil {
+		b.Settings["filetype"] = b.SyntaxDef.FileType
+	} else {
+		// search for the default file in the user's custom syntax files
+		b.SyntaxDef = findRealRuntimeSyntaxDef("default", nil)
+		if b.SyntaxDef == nil {
+			// search for the default file in the built-in syntax files
+			b.SyntaxDef = findRuntimeSyntaxDef("default", nil)
+		}
+	}
+
 	if b.SyntaxDef != nil && highlight.HasIncludes(b.SyntaxDef) {
 		includes := highlight.GetIncludes(b.SyntaxDef)
 
 		var files []*highlight.File
+		// TODO: search in the user's custom syntax files first
```

Comment:
Yep, user defined includes aren't considered yet, but are worth to be considered.
It wasn't in the scope of my PR.

Reply:
Yeah, this issue is separate from your PR, this mess has been here before.

Reply:
Maybe the following patch is a point to start...
[RefactorIncludesInUpdateRules.patch.txt](https://github.com/zyedidia/micro/files/15045077/RefactorIncludesInUpdateRules.patch.txt)
...it's not yet tested, but maybe you can improve and extend it or you throw it away and create your own approach. :wink: 

It shall consider nested includes, as long as the include isn't present in the list.

Open TODOs:
- prevent cyclic include
- process user files as includes

Reply:
> It shall consider nested includes

Hmm yeah, good point, thanks.

Though it seems your patch will still not fix it: `highlight.ResolveIncludes()` is still gonna handle the current file only.

Looks like `ParseDef()` itself would need to take care about includes, transparently?

Reply:
> Though it seems your patch will still not fix it: `highlight.ResolveIncludes()` is still gonna handle the current file only.

Hm, the problem is that...
```
func resolveIncludesInDef(files []*File, d *Def) {
	for _, lang := range d.rules.includes {
		[...]
	}
        [...]
}
```
...in `resolveIncludesInDef()`, which again targets the base syntax definition and his level of includes only. Looks like we could drop this outer loop. Then it will append the (unfortunately) once again parsed rules to the base syntax definition, what I would expect it to do...since we already collected all files involved.

After looking once again at my updated version of `parseDefFromFile()` (three returns) I came to the conclusion, that this isn't the best approach, since every iterated file will be fully parsed (all steps), even when we don't need it. So I will return to the single parsing steps within the `searchSyntaxDefIncludes()` till we've a better way to optimize the duplication there.

> Looks like `ParseDef()` itself would need to take care about includes, transparently?

No bad idea, to fully keep it out of the `buffer`. But then we need to find a way how to deal with the different runtime files, controlled and feed by the `buffer`.

Reply:
...I was wrong: we already ensure that in the list returned by `config.ListRuntimeFiles()` user's custom files always come before default files. Moreover, since your #3066 we ensure that the same filetype never appears twice: if there are both a default file and a user's custom file for the given filetype, this list only contains the user's custom one. So we don't need to fix anything in this regard.

...Regarding nested includes: yes, jumping back and forth between the `buffer` and the `highlighter` seems to be a no-go, it would be tremendously ugly and would be just a work around the broken design of the highlight parser's API. The way to go seems to be to redesign this API.

But I don't feel like redesigning it right now, so maybe let's just document that nested includes are not supported for now.

> [RefactorIncludesInUpdateRules2.patch.txt](https://github.com/zyedidia/micro/files/15048249/RefactorIncludesInUpdateRules2.patch.txt)
Not really convincing. :(

Actually the basic idea of your patch, i.e. just move the bulk of the include logic to a separate function and then fix it there, seems reasonable. Although, as I've just found, there is basically nothing to fix, but still it seems a good idea to move it to a separate function, to make `UpdateRules()` even shorter.